aten/src/ATen/core/Tensor.cpp

```diff
@@ -92,9 +93,183 @@ const Tensor& Tensor::requires_grad_(bool _requires_grad) const {
+      return "narrow";
+    case ViewMeta::Type::narrow_Tensor:
+      return "narrow_Tensor";
+    case ViewMeta::Type::movedim_intlist:
+      return "movedim_intlist";
+    case ViewMeta::Type::movedim_int:
+      return "movedim_int";
+    case ViewMeta::Type::reshape_as:
+      return "reshape_as";
+    case ViewMeta::Type::detach:
+      return "detach";
+    case ViewMeta::Type::transpose_int:
+      return "transpose_int";
+    case ViewMeta::Type::transpose_Dimname:
+      return "transpose_Dimname";
+    case ViewMeta::Type::unsqueeze:
+      return "unsqueeze";
+    case ViewMeta::Type::_values:
+      return "_values";
+    case ViewMeta::Type::indices:
+      return "indices";
+    case ViewMeta::Type::view:
+      return "view";
+    case ViewMeta::Type::view_dtype:
+      return "view_dtype";
+      default:
+        return "UNKNOWN_VIEW_TYPE";
+  }
+}
+
+std::ostream& operator<<(std::ostream& str, ViewMeta::Type rhs) {
+  return str << ViewMetaTypeToString(rhs);
+}
+
+Alias::Alias(const at::Tensor& base) : base_(base.clone()) {
+  auto impl = dynamic_cast<at::FunctionalTensorImplBase*>(base_.unsafeGetTensorImpl());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(impl != nullptr);
+  // Note [Marking Alias Tensors]
+  // In the functionalization pass, calling a view operation on a tensor forks it off an Alias object
+  // which is shared between the original tensor and the view.
+  // The Alias object contains a base tensor, which is used to sync operations across the view tensors.
+  // Throughout the Alias' life cylc,e we need to perform various views and reshapes on the base alias tensor.
+  // We don't want that process to recursively fork off another Alias object though, so we explicitly mark
+  // the base tensor as an alias.
+  // Another option would have been to implement separate {view}_copy() kernels for every corresponding view op,
+  // which have fallthrough kernels registered in the functionalization pass.
+  impl->mark_as_alias();
+}
+
+const at::Tensor& Alias::base() const {
+  return base_;
+}
+
+// metas is taken by value on purpose - we want to copy the vector.
+void Alias::add_update(const at::Tensor& updated_val, std::vector<ViewMeta> metas) {
+  updates_.push_back({updated_val, metas});
+  generation_++;
+}
+
+void Alias::apply_update(const Update& update) {
+  // TODO: Handle the other important view ops. This requires computing their inverses.
+  at::Tensor t = update.new_val;
+  for(int i = update.view_metas.size()-1; i >= 0; --i) {
+    switch (update.view_metas[i].view_type) {
+      case ViewMeta::Type::view:
+          t = t.view(update.view_metas[i].source_size);
+          break;
+      case ViewMeta::Type::noOp:
+          break;
+      default:
+          TORCH_CHECK(false, "Other types are not supported yet.");
+    }
+  }
+  base_ = t.clone();
+}
+
+void Alias::SyncUpdateOperations() {
+  for (auto& update_data: updates_) {
+    apply_update(update_data);
+  }
+  updates_.clear();
+}
```

Comment:
Why do the updates have to be queued? Is there a reason why we can't apply the update directly after an in-place operation?